File: sandbox.go
package agentsdk

import (
	"fmt"
	"os"
	"path/filepath"
	"strings"
)

// Sandbox manages the application's sandbox directory (~/.app-name/) and its
// standard subdirectories (data, cache, locks, crash_dumps).
//
// The base directory defaults to ~/.<app-name>/ and can be overridden by
// setting the environment variable <APP_NAME>_HOME (uppercased app name,
// hyphens replaced with underscores). For example, an app named "my-tool"
// would respect the MY_TOOL_HOME environment variable.
type Sandbox struct {
	appName  string
	baseDir  string
	subdirs  []string
}

// Standard subdirectory names.
const (
	SubdirData       = "data"
	SubdirCache      = "cache"
	SubdirLocks      = "locks"
	SubdirCrashDumps = "crash_dumps"
)

// NewSandbox creates a Sandbox for the given application name.
// The base directory is determined by:
//  1. The environment variable <APP_NAME>_HOME (uppercased, hyphens→underscores)
//  2. Falling back to ~/.<app-name>/ using os.UserHomeDir()
func NewSandbox(appName string) *Sandbox {
	subdirs := []string{SubdirData, SubdirCache, SubdirLocks, SubdirCrashDumps}

	envVar := strings.ToUpper(strings.ReplaceAll(appName, "-", "_")) + "_HOME"
	if dir := os.Getenv(envVar); dir != "" {
		return &Sandbox{
			appName: appName,
			baseDir: dir,
			subdirs: subdirs,
		}
	}

	homeDir, _ := os.UserHomeDir()
	baseDir := filepath.Join(homeDir, "."+appName)

	return &Sandbox{
		appName: appName,
		baseDir: baseDir,
		subdirs: subdirs,
	}
}

// Ensure idempotently creates the base directory and all subdirectories
// using os.MkdirAll with 0755 permissions.
func (s *Sandbox) Ensure() error {
	if err := os.MkdirAll(s.baseDir, 0755); err != nil {
		return fmt.Errorf("sandbox: create base dir %q: %w", s.baseDir, err)
	}
	for _, sub := range s.subdirs {
		path := filepath.Join(s.baseDir, sub)
		if err := os.MkdirAll(path, 0755); err != nil {
			return fmt.Errorf("sandbox: create subdir %q: %w", path, err)
		}
	}
	return nil
}

// BaseDir returns the sandbox's base directory path.
func (s *Sandbox) BaseDir() string {
	return s.baseDir
}

// DataDir returns the data subdirectory path.
func (s *Sandbox) DataDir() string {
	return filepath.Join(s.baseDir, SubdirData)
}

// CacheDir returns the cache subdirectory path.
func (s *Sandbox) CacheDir() string {
	return filepath.Join(s.baseDir, SubdirCache)
}

// LocksDir returns the locks subdirectory path.
func (s *Sandbox) LocksDir() string {
	return filepath.Join(s.baseDir, SubdirLocks)
}

// CrashDumpsDir returns the crash_dumps subdirectory path.
func (s *Sandbox) CrashDumpsDir() string {
	return filepath.Join(s.baseDir, SubdirCrashDumps)
}

// Dirs returns a map of subdirectory names to their full paths.
// Useful for agent doctor checks and diagnostic tooling.
func (s *Sandbox) Dirs() map[string]string {
	return map[string]string{
		SubdirData:       s.DataDir(),
		SubdirCache:      s.CacheDir(),
		SubdirLocks:      s.LocksDir(),
		SubdirCrashDumps: s.CrashDumpsDir(),
	}
}

File: sdks/python/agentsdk/sandbox.py
"""Sandbox — application directory manager.

Manages the application's sandbox directory (``~/.app-name/``) and its
standard subdirectories (data, cache, locks, crash_dumps).

The base directory defaults to ``~/.<app-name>/`` and can be overridden by
setting the environment variable ``<APP_NAME>_HOME`` (uppercased app name,
hyphens replaced with underscores). For example, an app named ``my-tool``
would respect the ``MY_TOOL_HOME`` environment variable.
"""

from __future__ import annotations

import os
from pathlib import Path
from typing import Dict

# Standard subdirectory names.
SUBDIR_DATA = "data"
SUBDIR_CACHE = "cache"
SUBDIR_LOCKS = "locks"
SUBDIR_CRASH_DUMPS = "crash_dumps"

_STANDARD_SUBDIRS = [SUBDIR_DATA, SUBDIR_CACHE, SUBDIR_LOCKS, SUBDIR_CRASH_DUMPS]


class Sandbox:
    """Manages the application's sandbox directory and its standard subdirectories."""

    def __init__(self, app_name: str) -> None:
        self._app_name = app_name
        self._subdirs = list(_STANDARD_SUBDIRS)

        # Check for environment variable override.
        env_var = app_name.upper().replace("-", "_") + "_HOME"
        env_dir = os.environ.get(env_var)
        if env_dir:
            self._base_dir = env_dir
        else:
            self._base_dir = str(Path.home() / f".{app_name}")

    # -- Accessors -----------------------------------------------------------

    @property
    def base_dir(self) -> str:
        """Return the sandbox's base directory path."""
        return self._base_dir

    @property
    def data_dir(self) -> str:
        """Return the data subdirectory path."""
        return str(Path(self._base_dir) / SUBDIR_DATA)

    @property
    def cache_dir(self) -> str:
        """Return the cache subdirectory path."""
        return str(Path(self._base_dir) / SUBDIR_CACHE)

    @property
    def locks_dir(self) -> str:
        """Return the locks subdirectory path."""
        return str(Path(self._base_dir) / SUBDIR_LOCKS)

    @property
    def crash_dumps_dir(self) -> str:
        """Return the crash_dumps subdirectory path."""
        return str(Path(self._base_dir) / SUBDIR_CRASH_DUMPS)

    def dirs(self) -> Dict[str, str]:
        """Return a map of subdirectory names to their full paths.

        Useful for agent doctor checks and diagnostic tooling.
        """
        return {name: str(Path(self._base_dir) / name) for name in self._subdirs}

    # -- Mutation ------------------------------------------------------------

    def ensure(self) -> None:
        """Idempotently create the base directory and all subdirectories.

        Raises:
            OSError: If the directory cannot be created (e.g. a file exists
                where a directory is expected).
        """
        base = Path(self._base_dir)
        try:
            base.mkdir(parents=True, exist_ok=True)
        except OSError as exc:
            raise OSError(f"sandbox: create base dir {self._base_dir!r}: {exc}") from exc

        for name in self._subdirs:
            sub_path = base / name
            try:
                sub_path.mkdir(parents=True, exist_ok=True)
            except OSError as exc:
                raise OSError(f"sandbox: create subdir {str(sub_path)!r}: {exc}") from exc
